### PR TITLE
[CDAP-13425] Add endpoints for saving and deleting reports

### PR DESCRIPTION
**Pull request by ecfm**
https://issues.cask.co/browse/CDAP-13425
https://builds.cask.co/browse/CDAP-DUT6441-3

The PR is separated in two commits:
1. Add fields in existing proto classes and proto/ReportGenerationRequest for the endpoints
2. Add endpoints and the logic for saving/deleting reports, also with unit test

### CODE REVIEW

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 1]
```diff
@@ -69,13 +70,16 @@
...
 import javax.annotation.Nullable;
+import javax.ws.rs.DELETE;
 import javax.ws.rs.DefaultValue;
 import javax.ws.rs.GET;
 import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 import javax.ws.rs.PathParam;
 import javax.ws.rs.QueryParam;
 
+import static co.cask.cdap.report.util.Constants.LocationName.SUCCESS_FILE;
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: cant you just import constants and reference success file ?

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 2]
```diff
@@ -137,17 +158,14 @@ public void getReports(HttpServiceRequest request, HttpServiceResponder responde
...
-          reportStatuses.add(new ReportStatusInfo(reportId, getReportMetaInfo(reportId, reportIdDir,
-                                                                              getReportRequest(reportIdDir))));
-        } catch (Exception e) {
-          LOG.error("Failed to get the status for report with id {}", reportId, e);
-          responder.sendError(500, String.format("Failed to get the status for report with id %s" +
-                                                   " because of error: %s", reportId, e.getMessage()));
-          return;
+        ReportMetaInfo metaInfo = getReportMetaInfo(reportId, reportIdDir, getReportRequest(reportIdDir));
+        // skip adding reports that are to be deleted
+        if (ReportStatus.DELETED.equals(metaInfo.getStatus())) {
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: if its not yet deleted, will be better to use the status `expired`

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 3]
```diff
@@ -137,17 +158,14 @@ public void getReports(HttpServiceRequest request, HttpServiceResponder responde
...
+          continue;
         }
+        reportStatuses.add(new ReportStatusInfo(reportId, metaInfo));
       }
-      responder.sendJson(200, new ReportList(offset, limit, reportIdDirs.size(), reportStatuses));
+      return new ReportList(offset, limit, reportIdDirs.size(), reportStatuses);
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java:177]: using offset blindly might cause duplicate results to be sent. lets say initially offset is 0 and limit is 10, but say report 1-5 is expired and you return the reports from 6-15 in the response. if the user will send the next query with offset 10 and limit 10, the reports from 11-15 will also be sent. which is incorrect, the correct request should have offset as 16, which can be known from previous query if the response had `idx` information.  another option is to improve the logic. where offset means skip non-expired records. offset of 10 would mean skip first 10 non-expired records, which i think is better. but this issue has to be addressed.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 4]
```diff
@@ -179,11 +197,15 @@ public void getReportStatus(HttpServiceRequest request, HttpServiceResponder res
...
       } catch (Exception e) {
-        LOG.error("Failed to get report generation info for report with id {}.", reportId, e);
-        responder.sendError(500, String.format("Failed to get report generation info for report with id %s" +
+        LOG.error("Failed to get the status for report with id {}.", reportId, e);
+        responder.sendError(500, String.format("Failed to get the status for report with id %s" +
                                                  " because of error: %s", reportId, e.getMessage()));
         return;
       }
+      if (ReportStatus.DELETED.equals(reportGenerationInfo.getStatus())) {
+        responder.sendError(404, String.format("Report with id %s does not exist.", reportId));
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: rather say, it has expired.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 5]
```diff
@@ -466,22 +612,42 @@ private void generateReport(ReportGenerationRequest reportRequest, Location repo
...
-     * Returns the status of the report generation by checking the presence of the success file or the failure file.
-     * If neither of these files exists, the report generation is still running.
+     * Gets the expiry time in milliseconds of a report by adding the VALID_DURATION_MILLIS
+     * to the report generation completed time read from the _SUCCESS file.
+     *
+     * @param reportIdDir the base directory with report ID as directory name
+     * @return the expiry time in milliseconds or {@code null} if the expiry time cannot be determined
+     * @throws Exception _SUCCESS doesn't exist or reading the report generation completed time from it failed
+     */
+    private static Long getExpiryTimeMillis(Location reportIdDir) throws Exception {
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java:695]: its better to keep the time unit as seconds. as the caller is converting it to seconds anyway.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 6]
```diff
@@ -241,10 +263,123 @@ private static ReportGenerationRequest getReportRequest(Location reportIdDir) th
     }
 
     private static ReportMetaInfo getReportMetaInfo(String reportId, Location reportIdDir,
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: add javadoc, since it has signification logic.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 7]
```diff
@@ -241,10 +263,123 @@ private static ReportGenerationRequest getReportRequest(Location reportIdDir) th
...
+        return;
+      }
+      // if the report is already deleted, deleting the report again is not allowed
+      if (ReportStatus.DELETED.equals(status)) {
+        responder.sendError(404, String.format("Report with id %s has already been deleted", reportId));
+        return;
+      }
+      Location deletedFile = reportIdDir.append(TO_BE_DELETED_FILE);
+      if (!deletedFile.createNew()) {
+        reportIdDir.delete();
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: why are you deleting the report dir if you cant create deleted file. that doesn't sound correct.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 8]
```diff
@@ -241,10 +263,123 @@ private static ReportGenerationRequest getReportRequest(Location reportIdDir) th
...
+        responder.sendError(500, "Failed to delete " + reportId);
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: error message should rather be, failed to complete delete due to {error-message}, please retry.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 9]
```diff
@@ -241,10 +263,123 @@ private static ReportGenerationRequest getReportRequest(Location reportIdDir) th
...
+        return;
+      }
+      // Save the time of when this report is marked as to-be-deleted in the _TO_BE_DELETED_FILE file
+      try (PrintWriter writer = new PrintWriter(new OutputStreamWriter(deletedFile.getOutputStream(),
+                                                                       StandardCharsets.UTF_8), true)) {
+        writer.write(Long.toString(System.currentTimeMillis()));
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: so on delete we don't delete the report but only mark it for deletion. that doesn't sound correct. Design doc mentions that it will delete the report. code needs to be fixed.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 10]
```diff
@@ -241,10 +263,123 @@ private static ReportGenerationRequest getReportRequest(Location reportIdDir) th
...
+          responder.sendError(404, String.format("Report with id %s does not exist.", reportId));
+          return;
+        default:
+          responder.sendError(403, "Cannot save the report with status " + status);
+          return;
+      }
+      // if the report is already saved, saving the report again is not allowed
+      Location savedFile = reportIdDir.append(SAVED_FILE);
+      if (savedFile.exists()) {
+        responder.sendError(403, String.format("Report with id %s is already saved and cannot be saved again.",
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: so we don't support updating description is that correct?

Comment (gsps1) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: i think we should support overwrite, there could be case where file is created and during write it fails and user wants to retry.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/proto/ReportStatus.java [hunk 11]
```diff
@@ -22,5 +22,6 @@
 public enum ReportStatus {
   RUNNING,
   COMPLETED,
-  FAILED
+  FAILED,
+  DELETED // only for internal use, will not be returned in the REST response as report generation status
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/proto/ReportStatus.java]: comment should be above.

File: cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala [hunk 12]
```diff
@@ -136,20 +136,35 @@ object ReportGenerationHelper {
     // TODO: [CDAP-13291] improve how the number of partitions is configured
     resultDf.coalesce(1).write.option("timestampFormat", "yyyy/MM/dd HH:mm:ss ZZ").json(reportDir)
     val count = resultDf.count
-    // Write the total number of records in _SUCCESS file generated after successful report generation
+    // Create a _COUNT file and write the total number of report records in it
+    writeToFile(count.toString, Constants.LocationName.COUNT_FILE, reportIdDir)
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala:140]: why is this needed?

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 3 above)
Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java:177]: right, I think the second approach is better

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 8 above)
Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: there is no exception thrown when `deletedFile.createNew()` returns false, from the implementation, it looks like it only happens when the file already exists, but it's not documented in the method's javadoc.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 9 above)
Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: it is marked for deletion but not immediately deleted because I want to avoid the possible conflict where the report is being read but is deleted at the same time. So I think it's better to mark it as expired and then let it to be deleted together with other expired reports after a certain period of time.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 10 above)
Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: the UI doesn't allow users to change the name and description once they save the report. For the case where the file is created but write operation failed, should we only allow updating the file if the content in the file is not a valid save request?

File: cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala (hunk 12 above)
Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala:140]: this is for providing the total number of report records in the response of reading the report details without the need of iterating through the whole report file

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 4 above)
Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: I think in a user's perspective, an expired report and a deleted report are essentially the same, since both cannot be listed, read, or checked for status, and both cannot be be brought back. Also the a report with status can also mean that the report is deleted by the user, so it will be confusing to say that this report has expired

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 5 above)
Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java:695]: there's also another usage of this method to compare the time returned directly with `System.currentTimeMillis()`

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 9 above)
Comment (gsps1) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: how will you avoid the conflict then ? that conflict can happen any time you delete it. there is no point in not deleting it now. say we haven't added the expiration currently and users want to delete old reports, using this endpoint/UI to do that is much better than asking them to delete the directories manually. i think we should immediately delete the directory for this REST endpoint as mentioned in design.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 10 above)
Comment (gsps1) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: implementation and what UI does are two separate things. if you dont want to allow updates to name/description, maybe  check if file exists and also able to get those details. and say updates to name/description is not supported.

Comment (maochf) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: if the report is marked as DELETED, users won't see them in the UI, unless they keep a note of the report ID and call the REST API directly. It provides a grace period if the user is already reading the report while the report is being deleted, especially when we support sharing reports among users.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 9 above)
Comment (chtyim) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: How large is a report? When reading the report, do you stream the report back or buffer in memory and then send back? If the report is relatively small and the deletion is fast, deleting in place seems make sense.

Generally we only do async deletion when deletion itself is complicated (e.g. it involves propagating the deletion to other related entities).

Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: reports in GB's will be considered large. currently we buffer the report in memory and send it back, but can improve it later. I'll make the deletion in place considering the conflict scenario is rare and won't actually cause much problem. This will simplify the current logic too

Comment (chtyim) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: Reports in GB? That's seems really huge. What's inside?

Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: just an extreme case like there are records with average size 400B, and 100 new records every minute, and the report is over a year. Normally the report should just be several MB's at most

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java [hunk 13]
```diff
@@ -637,15 +710,15 @@ private static Long getExpiryTimeMillis(Location reportIdDir) throws Exception {
      * @return status of the report generation
      */
     private static ReportStatus getReportStatus(Location reportIdDir) throws Exception {
-      if (reportIdDir.append(TO_BE_DELETED_FILE).exists() ||
-        // if the report is completed but has expired, return DELETED as its status too
-        (reportIdDir.append(SUCCESS_FILE).exists() && getExpiryTimeMillis(reportIdDir) < System.currentTimeMillis())) {
-        return ReportStatus.DELETED;
+      // if the report is completed but has expired, return EXPIRED as its status too
+      if (reportIdDir.append(LocationName.SUCCESS_FILE).exists() &&
+          getExpiryTimeMillis(reportIdDir) < System.currentTimeMillis()) {
```
Comment (gsps1) [inline cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: why extra space ?

**Review by gsps1 — COMMENTED** [review]
minor comment on indentation, can you add TODO and JIRA on adding a scheduled service  to delete the expired reports. it should be simple enough to be addressed in this PR if you want too. lgtm otherwise.

File: cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java (hunk 13 above)
Comment (ecfm) [reply cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java]: removed

**Comment by ecfm** [comment]
build is green. merging

**Comment by ecfm** [comment]
There is an existing JIRA for cleaning up the expired files https://issues.cask.co/browse/CDAP-13427
